transport.go

```diff
@@ -32,27 +32,72 @@ var quicConfig = &quic.Config{
 }
 
 type connManager struct {
-	connIPv4Once sync.Once
-	connIPv4     net.PacketConn
+	mu          sync.Mutex
+	defaultConn net.PacketConn
+	// map underhood PacketConn -> (re)use count
+	ipv4Conns map[net.PacketConn]int
+	ipv6Conns map[net.PacketConn]int
+}
 
-	connIPv6Once sync.Once
-	connIPv6     net.PacketConn
+func newConnManager() *connManager {
+	return &connManager{
+		ipv4Conns: make(map[net.PacketConn]int),
+		ipv6Conns: make(map[net.PacketConn]int),
+	}
 }
 
-func (c *connManager) GetConnForAddr(network string) (net.PacketConn, error) {
+func (c *connManager) pickDialConn(conns map[net.PacketConn]int, remoteHostIP net.IP) net.PacketConn {
+	for conn := range conns {
+		addr := conn.LocalAddr().(*net.UDPAddr)
+		if addr.IP.IsLoopback() {
+			if remoteHostIP.IsGlobalUnicast() {
+				continue
+			}
+			if remoteHostIP.IsLoopback() {
+				return conn
+			}
+		}
+		if c.defaultConn == nil && (addr.IP.IsGlobalUnicast() || addr.IP.IsUnspecified()) {
+			c.defaultConn = conn
+		}
+		if addr.IP.IsUnspecified() && remoteHostIP.IsGlobalUnicast() {
+			return conn
+		}
+	}
+	return c.defaultConn
+}
+
+func (c *connManager) GetConnForAddr(network, host string) (net.PacketConn, error) {
+	remoteAddr, err := net.ResolveUDPAddr(network, host)
+	if err != nil {
+		return nil, err
+	}
+
 	switch network {
 	case "udp4":
-		var err error
-		c.connIPv4Once.Do(func() {
-			c.connIPv4, err = c.createConn(network, "0.0.0.0:0")
-		})
-		return c.connIPv4, err
+		conn := c.pickDialConn(c.ipv4Conns, remoteAddr.IP)
```

Comment:
This is going to need some synchronization (we're reading a shared ipv4Conns map and writing a shared `c.defaultConn` variable.